holoviews/operation/downsample.py

```diff
@@ -176,7 +183,33 @@ def _process(self, element, key=None):
             return element.clone(elements)
 
         if self.p.x_range:
-            element = element[slice(*self.p.x_range)]
+            try:
+                mask = element.dataset.interface._select_mask_neighbor(
+                    element.dataset, {element.kdims[0]: self.p.x_range}
+                )
+            except NotImplementedError:
+                mask = slice(*self.p.x_range)
+            except Exception as e:
+                self.param.warning(f"Could not apply neighbor mask to downsample1d: {e}")
+                mask = slice(*self.p.x_range)
+            element = element[mask]
+        elif self.p.algorithm == "viewport-xlim":
+            # We only want to send some of the data to the browser on the
+            # first pass if we have set xlim. This will render the plot and
+            # trigger a second pass with the x_range set. This will not work
+            # with the matplotlib backend because it does not update the x_range
+            # after the first pass, but that is not a problem because the matplotlib
+            # backend with the viewport algorithm does not make any sense.
+            # This is not very elegant.
+            size = element.dataset.shape[0]
+            mask1 = element.dataset.interface.select_mask(
+                element.dataset, {element.kdims[0]: slice(self.p.width // 2)},
+            )
+            mask2 = element.dataset.interface.select_mask(
+                element.dataset, {element.kdims[0]: slice(size - self.p.width // 2, size)},
+            )
+            element = element[mask1 | mask2]
```

Comment:
An alternative is to switch to `nth`  downsample when `self.p.algorithm == "viewport-xlim" and self.p.x_range is None`.